DM-22677: Relocate scripts to python module

timj opened the pull request.

timj committed:
Relocate script logic to python module

timj committed:
Add simple test for new script module

timj committed:
Add scripts to documentation

timj committed:
Change Note to Notes in numpydoc

TallJimbo commented on python/lsst/daf/butler/script/dumpButlerConfig.py:
```diff
@@ -0,0 +1,100 @@
+        print(config, file=outfh)
+
+
+def main():
```
Would it be worthwhile having this take an `args` to pass to `parse_args()` that would take `sys.argv[1:]` or similar?  That seems like it's a missing piece of making this more usable, but I also don't really a scenario where we'd actually use that.

timj replied:
I believe I'm following standard practice here: https://developer.lsst.io/stack/argparse-script-topic-type.html?highlight=autoprogram

TallJimbo commented on python/lsst/daf/butler/script/dumpButlerConfig.py:
```diff
@@ -0,0 +1,100 @@
+    try:
+        dumpButlerConfig(args.root, args.searchpath, args.subset, sys.stdout)
+    except Exception as e:
+        print(f"{e}", file=sys.stderr)
```
I'm not sure I like this traceback squashing here. In a really mature program that knows about all of its common failure modes, that can give you more friendly error messages, but in code that's under development, so often it just hides what went wrong and forces you to run again, which I find super frustrating.

timj replied:
I got so wrapped up in ensuring that I returned non-zero exit status that I forgot I could just let it die and also get non-zero exit status.

TallJimbo commented on python/lsst/daf/butler/core/repoTransfers.py:
```diff
@@ -165,8 +165,8 @@ def saveDatasets(self, refs: Iterable[DatasetRef], *,
-        Note
-        ----
+        Notes
+        -----
```
I think I've also got this fix on a branch for you to review :slightly_smiling_face: .

TallJimbo reviewed (approved):
Two minor comments on one script that probably apply to all of them.

timj committed:
For scripts do not catch exceptions

Exceptions lead to bad exit status anyhow so there is no need
to catch the exception just to return a bad exit status.
